lib/ansible/module_utils/facts/facts.py

```diff
@@ -0,0 +1,69 @@
+#       of a list of the fact gather classes to include (finer grained gather_facts)
+#       Or, possibly even a list or dict of fact labels 'ansible_lvm' for ex, that
+#       the driver class would use to determine which fact gathering classes to load
+class Facts:
+    """
+    This class should only attempt to populate those facts that
+    are mostly generic to all systems.  This includes platform facts,
+    service facts (e.g. ssh keys or selinux), and distribution facts.
+    Anything that requires extensive code or may have more than one
+    possible implementation to establish facts for a given topic should
+    subclass Facts.
+    """
+
+    # For the most part, we assume that platform.dist() will tell the truth.
+    # This is the fallback to handle unknowns or exceptions
+
+    # NOTE: load_on_init is changed for ohai/facter classes. Ideally, all facts
+    #       would be load_on_init=False and this could be removed. -akl
+    # NOTE: cached_facts seems like a misnomer. Seems to be used more like an accumulator -akl
+    # nothing should be using cached_facts args now, so removing it
+    def __init__(self, module, load_on_init=True):
+
+        self.module = module
+
+        # FIXME: This is where Facts() should end, with the rest being left to some
+        #        composed fact gathering classes.
+
+        # TODO: Eventually, these should all get moved to populate().  But
+        # some of the values are currently being used by other subclasses (for
+        # instance, os_family and distribution).  Have to sort out what to do
+        # about those first.
+        # NOTE: if the various gathering methods take a arg that is the 'accumulated' facts
+        #       then this wouldn't need to happen on init. There would still be some ordering required
+        #       though. If the gather methods return a dict of the new facts, then the accumulated facts
+        #       can be read-only to avoid manipulating it by side effect. -akl
+
+        # TODO: to avoid hard coding this, something like
+        # list so we can imply some order suggestions
+        # fact_providers is a map or lookup of fact label -> fact gather class/inst that provides it
+        #  - likely will also involve a fact plugin lookup
+        #    ( could fact providing modules include the list of fact labels in their metadata? so we could determine
+        #      with plugin to load before we actually load and inst it?)
+        # fact_gatherers = []
+        # for requested_fact in requested_facts:
+        #    fact_gatherer = self.fact_providers.get('requested_fact', None)
+        #    if not fact_gatherer:
+        #        continue
+        #    fact_gatherers.append(fact_gatherer)
+
+        # TODO: de-dup fact_gatherers
+        # for gatherer in fact_gatherers:
+        #    data = gatherer.gather()
+        #    self.facts.update(data)
+
+    def populate(self):
+        return self.facts
```

Comment:
We could consider using abc's for this aswell.  Now that we're python-2.6+ we are free to make use of those moduleside.  (You can look at the connection plugin __init__.py for an example if you haven't used them before).  One caution about this idea is that I'm not sure if abcs and __new__ will play together transparently.  In my own code, I've had to do extra work to get metaclasses and abcs to work together (since abc's are a metaclass) but I'm not sure if __new__ is sufficiently different from a full-blown metaclass that it works out of the box.